### ISSUE TITLE
TreeSet: RemoveFunc not remove all  matched node 

### ISSUE THREAD

**Issue by zonewave** [hashicorp/go-set#63]
for example, I want to remove all multiples of 3，
```golang

func ExampleTreeSet_RemoveFunc() {
	s := TreeSetFrom[int, Compare[int]](ints(20), Cmp[int])

	fmt.Println(s)

	even := func(i int) bool {
		return i%3 != 0
	}
	s.RemoveFunc(even)

	fmt.Println(s)

	// Output:
	// [1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20]
	// [3 6 9 12 15 18 20]
}

```
but  run test:
![image](https://github.com/hashicorp/go-set/assets/26363539/e51095db-c3f5-4a41-8a65-2dd315c13b4c)

I think It is  during the deletion walk, the structure of the tree changes so that some nodes cannot be visisted.

The simple way to solve the problem :
1. first ,find all matched removeIds
2. second, use RemoveSlice  to remove
 ```golang
func (s *TreeSet[T, C]) RemoveFunc(f func(T) bool) bool {
	modified := false
	removeIds := make([]T, 0)
	findRemoveIds := func(item T) bool {
		if f(item) {
			removeIds = append(removeIds, item)
		}
		return true
	}
	s.ForEach(findRemoveIds)
	if len(removeIds) > 0 {
		modified = true
	}
	s.RemoveSlice(removeIds)
	return modified
}
```

**Comment by shoenig** [comment]
Oof yeah that's a legitimate bug @zonewave!

In fact I [tried to document](https://github.com/hashicorp/go-set/blob/main/treeset.go#L45) that modifying the tree while traversing it cannot work... and then forgot :sweat:. I remember the Java standard library tries to detect when this happens and throws [ConcurrentModificationException](https://docs.oracle.com/javase/8/docs/api/java/util/ConcurrentModificationException.html) ... but I think for now we can start with just fixing our own bug.

Can you open a PR? I think you've got the right idea.

**Comment by shoenig** [comment]
Closed by #64
